variables.tf

```diff
@@ -53,20 +53,21 @@ variable "sg_ingress_rules" {
 }
 
 variable "subnet_ids" {
+  description = "Target Subnet IDs for \"Interface\" services. Also used to resolve the `vpc_id` for \"Gateway\" services"
   type        = list(string)
-  description = "Target Subnet ids."
-  default     = []
 }
 
 variable "vpc_endpoint_services" {
-  type        = list(string)
-  description = "List of AWS Endpoint service names that are used to create VPC Interface Endpoints. Both Gateway and Interface Endpoints are supported. See https://docs.aws.amazon.com/general/latest/gr/rande.html for full list."
-  default     = []
+  description = "List of AWS Endpoint service names and types. Both Gateway and Interface Endpoints are supported. See https://docs.aws.amazon.com/general/latest/gr/rande.html for full list."
+  type = list(object({
```

Comment:
🤔 I think a `list of maps` makes things terribly verbose/redundant as compared to the two lists.

I do generally like data structures as input vars though. What if the variable is defined as:
```hcl
variable "vpc_endpoint_services" {
  type = object({
    interface = list(string)
    gateway   = list(string)
  })
  default = {
    interface = []
    gateway   = []
  }
}
```

Then the input from the user looks like:
```hcl
module "foo" {
  vpc_endpoint_services = {
    interface = ["sns", "sqs", "config"]
    gateway   = ["s3", "dynamo"]
  }
}
```

Note that I'm typing the object declaration from my memory of tf syntax .. it could be wrong .. or just plain not be possible. But I wanted to pose the idea.

Reply:
You're not wrong on the verbosity, but I was also thinking about using that data structure in the future to do other interesting things... e.g. Have an externally-managed list of security groups you want to use with just this one endpoint? Ok, expose `security_groups` as another attribute.

Also, once the "optional" feature for object attributes goes live, we can also easily implement defaults, to make things much less verbose.

Reply:
To add some context to where my `verbose/redundant` comment comes from: our "default" input var declaration will go from `22` lines to `85` lines. It's a tad of a nitpick, but at the moment we're only enabling `20` out of a possible ~`75` endpoints! If we chose to enable all of them we'd have to declare a variable at near `300` lines!

Reply:
Is that all in a tfvars file you use with this module directly? Or do you have a .tf file where you can construct that value using a for loop? The interface *you* expose to this module doesn't have to change, necessarily...

Reply:
Just pushed an example of the new `optional` attribute feature, currently only available as an experiment. But this is what I'm thinking to reduce a lot of that verbosity...

Reply:
Thanks for the updates!

Our implementation is indeed behind an interface. We use this module in a couple of different ways and neither requires a user to "configure" it. One location wraps this in another module strictly for sharing a common set of enabled endpoints -- include the wrapper, no config passed, done. The second location is more opinionated about its endpoints but hides away the details from the user (automated, declarative, internal Python tool). So, in neither case is the verbosity a huge issue for us.

At the moment it still feels a tad on the edge of adding complexity for little to no benefit. But! The idea of being able to add additional config per-endpoint-service is interesting and requires such an object interface to make it not-ugly. I'm onboard I think.

---

A side note on the optional attribute: It _requires_ a tf 0.14 upgrade. That shouldn't be a big issue for anyone already on the version requiring 0.12. Though it will likely come as a surprise for many. The bigger question on my mind is if the _experimental_ feature should be used in a production module? I've been leary of Terraform experiments in the past. I think I am in favor of verbosity over experimental features. 😀 

If the `type` attribute is not optional you can stick with a minimum tf version of 0.12 but it'll be compatible all the way to 0.14. This is, to me, yet another win for verbosity.

Reply:
> The bigger question on my mind is if the experimental feature should be used in a production module?

I wouldn't implement the `optional()` modifier until that feature is released GA. It was just meant as an example of what it _could_ look like, in a future version.

> I'm onboard I think

I was chatting with another dev internally, and we're both favoring this typed object. So I'll clean up this branch and we can get this released!

Reply:
Sounds good! Thanks for the considerate back and forth; it's much appreciated.